File: app/src/main/java/fr/nihilus/music/utils/ViewUtils.java
package fr.nihilus.music.utils;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Color;
import android.os.Build;
import android.support.annotation.AttrRes;
import android.support.annotation.ColorInt;
import android.support.annotation.FloatRange;
import android.support.annotation.NonNull;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.View;

public final class ViewUtils {

    public static void setLightStatusBar(@NonNull View view, boolean darkIcons) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            view.setSystemUiVisibility(darkIcons ? View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR : 0);
        }
    }

    /**
     * Indique si une couleur est claire.
     * Cette définition s'appuie sur le niveau de saturation et la luminosité.
     *
     * @param color couleur à tester
     * @return true si la couleur est claire, false si elle est foncée
     */
    public static boolean isColorBright(@ColorInt int color) {
        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        return (hsv[1] < 0.5) && (hsv[2] > 0.5);
    }

    @ColorInt
    public static int darker(@ColorInt int color, @FloatRange(from = 0.0, to = 1.0) float factor) {
        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        hsv[2] *= factor;
        return Color.HSVToColor(hsv);
    }

    /**
     * Donne le nombre de pixels équivalent à un nombre de DIP (Density-Independant Pixel).
     * Cette unité permet d'obtenir des dimensions similaires sur des écrans de taille
     * et de densité différentes.
     *
     * @param dp nombre de dp à convertir en pixels
     * @return nombre de pixels
     */
    public static int dipToPixels(Context context, float dp) {
        final DisplayMetrics metrics = context.getResources().getDisplayMetrics();
        return Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, metrics));
    }

    /**
     * Récupère une couleur appartenant au thème de l'application, telle que "colorPrimary"
     * ou "colorAccent".
     *
     * @param themeAttr attribut du thème pointant sur la couleur
     * @return nombre entier représentant cette couleur
     */
    @ColorInt
    public static int resolveThemeColor(Context context, @AttrRes int themeAttr) {
        TypedValue outValue = new TypedValue();
        Resources.Theme theme = context.getTheme();
        theme.resolveAttribute(themeAttr, outValue, true);
        return outValue.data;
    }
}

File: app/src/main/java/fr/nihilus/music/utils/ViewUtils.kt
/*
 * Copyright 2017 Thibault Seisel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:JvmName("ViewUtils")

package fr.nihilus.music.utils

import android.content.Context
import android.graphics.Color
import android.support.annotation.AttrRes
import android.support.annotation.ColorInt
import android.support.annotation.FloatRange
import android.support.annotation.Px
import android.util.TypedValue

@Px
fun dipToPixels(context: Context, dp: Float): Int {
    val metrics = context.resources.displayMetrics
    return Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, metrics))
}

/**
 * Produce a darker shade of this color by a given factor.
 */
@ColorInt
fun darker(@ColorInt color: Int, @FloatRange(from = 0.0, to = 1.0) factor: Float): Int {
    val hsv = FloatArray(3)
    Color.colorToHSV(color, hsv)
    hsv[2] *= factor
    return Color.HSVToColor(hsv)
}

/**
 * Resolve a color attribute from the application theme such as `colorPrimary` or `colorAccent`.
 *
 * @param themeAttr theme attribute pointing on that color
 * @return the desired color packed as an aRGB color int.
 */
@ColorInt
fun resolveThemeColor(context: Context, @AttrRes themeAttr: Int): Int {
    val outValue = TypedValue()
    val theme = context.theme
    theme.resolveAttribute(themeAttr, outValue, true)
    return outValue.data
}
